dynawo/sources/Models/CPP/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp

```diff
@@ -0,0 +1,579 @@
+        fLocal_[3] = Ii() - (P() * Ui() - Q() * Ur()) / (U() * U());
+        fLocal_[4] = P() - P0_ * pow(U() / Uf(), alpha_);
+        fLocal_[5] = Q() - Q0_ * pow(U() / Uf(), beta_);
+      }
+    } else {
+      fLocal_[0] = UfRaw();
+      fLocal_[1] = Uf();
+      fLocal_[2] = Ir();
+      fLocal_[3] = Ii();
+      fLocal_[4] = P();
+      fLocal_[5] = Q();
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::setFequations() {
+    if (getIsInitProcess())
+      return;
+    if (isConnected()) {
+      unsigned int index = 0;
+      fEquationIndex_[index] = std::string("Tf_*UfPrim() - (U() - Uf()):").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Uf() - UfRaw():").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Ir() - (P() * Ur() + Q() * Ui()) / (U() * U()): ").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Ii() - (P() * Ui() - Q() * Ur()) / (U() * U()): ").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("P() - P0_ * (U()/Uf())^alpha_ :").append(staticId());
+      index++;
+      fEquationIndex_[index] = std::string("Q() - P0_ * (U()/Uf())^beta_ :").append(staticId());
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalZ(const double & /*t*/) {
+    State connectionState1 = CLOSED;
+    if (zLocal_[0] > 0) {
+      connectionState1 = OPEN;
+    }
+    State connectionState2 = CLOSED;
+    if (zLocal_[1] > 0) {
+      connectionState2 = OPEN;
+    }
+    if (connectionState1 != connectionState1_prev) {
+      setConnected(connectionState1);
+      connectionState1_prev = connectionState1;
+    }
+    if (connectionState2 != connectionState2_prev) {
+      setConnected(connectionState2);
+      connectionState2_prev = connectionState2;
+    }
+    if (gLocal_[0] == ROOT_UP) {
+      UMaxPuReached_ = true;
+      UMinPuReached_ = false;
+      return;
+    }
+    if (gLocal_[1] == ROOT_UP) {
+      UMaxPuReached_ = false;
+      UMinPuReached_ = true;
+      return;
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalJt(const double& /*t*/, const double& cj, SparseMatrix& jt, const int& rowOffset) {
+    if (getIsInitProcess())
+      return;
+    if (!isConnected()) {
+      jt.changeCol();
+      jt.addTerm(UfRawYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(UfYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(IrYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(IiYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(PYNum_ + rowOffset, 1);
+      jt.changeCol();
+      jt.addTerm(QYNum_ + rowOffset, 1);
+    } else {
+      jt.changeCol();
+      jt.addTerm(UfRawYNum_ + rowOffset, 1.0 + cj * Tf_);
+      jt.addTerm(UrYNum_ + rowOffset, -Ur()/U());
+      jt.addTerm(UiYNum_ + rowOffset, -Ui()/U());
+      jt.changeCol();
+      jt.addTerm(UfYNum_ + rowOffset, 1);
+      if (UMaxPuReached_ == false && UMinPuReached_ == false) {
```

Comment:
This condition is not visible in the evalF

Reply:
the condition is expressed via the variable UfRawValue in evalF